lib/http.c

```diff
@@ -1919,14 +1918,14 @@ CURLcode Curl_http(struct connectdata *conn, bool *done)
 #ifdef HAVE_LIBZ
   /* we only consider transfer-encoding magic if libz support is built-in */
 
-  if(!Curl_checkheaders(conn, "TE:") &&
+  if(!Curl_checkheaders(conn, "TE") &&
      data->set.http_transfer_encoding) {
     /* When we are to insert a TE: header in the request, we must also insert
        TE in a Connection: header, so we need to merge the custom provided
        Connection: header and prevent the original to get sent. Note that if
        the user has inserted his/hers own TE: header we don't do this magic
        but then assume that the user will handle it all! */
-    char *cptr = Curl_checkheaders(conn, "Connection:");
+    char *cptr = Curl_checkheaders(conn, "Connection");
 #define TE_HEADER "TE: gzip\r\n"
 
```

Comment:
Before creating an updated "Connection" header, a check is necessary that the header is not empty, e.g.

```
char *cptr = Curl_checkheaders(conn, "Connection");
if(cptr && cptr[10] != ':')
  cptr = NULL;

if(cptr) {
  while(*cptr && ISSPACE(*cptr))
    cptr++;

  if (!*cptr)
    cptr = NULL;
}
```

Reply:
I'm not sure I follow.

For transfer-encoding it *must* send a `Connection:` header which is what this ensures. It will override the user's (conflicting) decision to remove or blank it, if there is one. This is actually a bug already in the current code, as passing -H `Connection:` and --tr-encoding will send the somewhat odd-looking header `Connection: ,TE"

Are you suggesting the order to remove or blank the `Connection:` header should be stronger than the request to ask for transfer-encoding? I'm proposing the other way around since a user can't get transfer-encoding without the header, and if a user really doesn't want the `Connection:` header it makes no sense to ask for transfer-encoding!

Reply:
> Are you suggesting the order to remove or blank the Connection: header should be stronger than the request to ask for transfer-encoding?

Yes, that was my intention. But now I think you're right, we should not change the existing behavior. So a "TE" should be added to "Connection", even if the user specified "Connection;" in the header list.

The existing code needs to be tweaked a bit to avoid adding a "Connection; TE" (with semicolon instead of colon).

Reply:
Agreed. I'll add a test case doing exactly that to make sure it gets right...